assets/js/interactivity/hooks.tsx

```diff
@@ -0,0 +1,299 @@
+ * ```
+ *
+ * This could be an hypothetical implementation of the custom directive used in
+ * the snippet above.
+ *
+ * @example
+ * ```js
+ * directive(
+ *   'color', // Name without prefix and suffix.
+ *   ( { directives: { color }, ref, evaluate }) => {
+ *     if ( color.text ) {
+ * 	     ref.style.setProperty(
+ *         'color',
+ *         evaluate( color.text )
+ *       );
+ *     }
+ *     if ( color.background ) {
+ *       ref.style.setProperty(
+ *         'background-color',
+ *         evaluate( color.background )
+ *       );
+ *     }
+ *   }
+ * )
+ * ```
+ *
+ * @param {string}            name     Directive name, without the `data-wp-` prefix.
+ * @param {DirectiveCallback} callback Function that runs the directive logic.
+ * @param {DirectiveOptions=} options  Options object.
+ */
+export const directive = ( name, callback, { priority = 10 } = {} ) => {
+	directiveCallbacks[ name ] = callback;
+	directivePriorities[ name ] = priority;
+};
+
+// Resolve the path to some property of the store object.
+const resolve = ( path, namespace ) => {
+	let current = {
+		...stores.get( namespace ),
+		context: getScope().context[ namespace ],
+	};
+	path.split( '.' ).forEach( ( p ) => ( current = current[ p ] ) );
+	return current;
+};
+
+// Generate the evaluate function.
+const getEvaluate =
+	( { scope } = {} ) =>
+	( entry, ...args ) => {
+		let { value: path, namespace } = entry;
+		// If path starts with !, remove it and save a flag.
+		const hasNegationOperator =
+			path[ 0 ] === '!' && !! ( path = path.slice( 1 ) );
+		setScope( scope );
+		const value = resolve( path, namespace );
+		const result = typeof value === 'function' ? value( ...args ) : value;
+		resetScope();
+		return hasNegationOperator ? ! result : result;
+	};
+
+// Separate directives by priority. The resulting array contains objects
+// of directives grouped by same priority, and sorted in ascending order.
+const getPriorityLevels = ( directives ) => {
+	const byPriority = Object.keys( directives ).reduce( ( obj, name ) => {
+		if ( directiveCallbacks[ name ] ) {
+			const priority = directivePriorities[ name ];
+			( obj[ priority ] = obj[ priority ] || [] ).push( name );
+		}
+		return obj;
+	}, {} );
+
+	return Object.entries( byPriority )
+		.sort( ( [ p1 ], [ p2 ] ) => p1 - p2 )
+		.map( ( [ , arr ] ) => arr );
+};
+
+// Priority level wrapper.
+const Directives = ( {
+	directives,
+	priorityLevels: [ currentPriorityLevel, ...nextPriorityLevels ],
+	element,
+	originalProps,
+	previousScope = {},
+} ) => {
+	// Initialize the scope of this element. These scopes are different per each
+	// level because each level has a different context, but they share the same
+	// element ref, state and props.
+	const scope = useRef( {} ).current;
+	scope.context = useContext( context );
+	scope.ref = previousScope.ref || useRef( null );
+	scope.state = previousScope.state || useRef( deepSignal( {} ) );
+	scope.props = element?.props || originalProps;
```

Comment:
Oh, right. The `element` prop is always defined. I'll update that.

PS: I've noticed that we're adding the props of the previous element instance (we're cloning the element for each priority level). I'm not sure if that could cause problems accessing newly added properties, but just to prevent any potential issue, we could clone `element` before adding `props` to the scope. 🤔

Reply:
Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/11071/commits/0662935488af86c982fbe16001a201b887ca8887.